Modules/_cursesmodule.c

```diff
@@ -181,32 +197,55 @@ static int curses_start_color_called = FALSE;
 static const char *curses_screen_encoding = NULL;
 
 /* Utility Macros */
-#define PyCursesSetupTermCalled                                         \
+
+/*
+ * Macro to check that FUNC_NAME has been called by testing
+ * the CALLED boolean. If an error occurs, a PyCursesError
+ * is raised.
+ *
+ * Since these macros can be called in functions that do not
+ * have a direct access to the module's state, the exception
+ * type is imported on demand as well.
+ */
+#define _PyCursesCheckFunction(CALLED, FUNC_NAME)                       \
     do {                                                                \
-        if (curses_setupterm_called != TRUE) {                          \
-            PyErr_SetString(PyCursesError,                              \
-                            "must call (at least) setupterm() first");  \
+        if ((CALLED) != TRUE) {                                         \
+            PyObject *exc = _PyImport_GetModuleAttrString("_curses",    \
+                                                          "error");     \
+            if (exc == NULL) {                                          \
+                return 0;                                               \
+            }                                                           \
+            PyErr_SetString(exc, "must call " # FUNC_NAME "() first");  \
+            Py_DECREF(exc);                                             \
             return 0;                                                   \
         }                                                               \
     } while (0)
 
-#define PyCursesInitialised                                 \
-    do {                                                    \
-        if (curses_initscr_called != TRUE) {                \
-            PyErr_SetString(PyCursesError,                  \
-                            "must call initscr() first");   \
-            return 0;                                       \
-        }                                                   \
+/*
+ * Macro to check that FUNC_NAME has been called by testing
+ * the CALLED boolean. If an error occurs, a PyCursesError
+ * is raised. The exception type is obtained from the module
+ * state.
+ */
+#define _PyCursesStatefulCheckFunction(CALLED, FUNC_NAME, MODULE)       \
```

Comment:
Can it become a static inline function?

Reply:
Not entirely sure since this one is usually used in the body of more complex functions that have something after. By making it an inline function, we would need to change calls of the form

```C
PyCursesStatefulInitialised(module);
PyCursesStatefulInitialisedColor(module);
...
```

into

```C
if (!PyCursesStatefulInitialised(module)) return NULL;
if (!PyCursesStatefulInitialisedColor(module)) return NULL;
```

but if you want me to do this I can also do it.

Reply:
Ok, found a way to do it.